modules/3d/src/spc.cpp

```diff
@@ -0,0 +1,244 @@
+        }
+    }
+    // init
+    Mat g_dist_mat = Mat::zeros(num_faces, num_faces, CV_32F);
+    Mat a_dist_mat = Mat::zeros(num_faces, num_faces, CV_32F);
+    std::vector<Point3f> f1, f2;
+    std::pair<Point3f, Point3f> edge;
+    // compute distances
+    for (const auto& entry : map) {
+        const std::pair<int, int>& adjacent_faces = entry.second;
+        if (adjacent_faces.second == -1)
+            continue;
+        ++num_adjacency;
+        const std::vector<int32_t>& face1 = indices[adjacent_faces.first];
+        const std::vector<int32_t>& face2 = indices[adjacent_faces.second];
+
+        f1 = {vertices[face1[0]], vertices[face1[1]], vertices[face1[2]]};
+        f2 = {vertices[face2[0]], vertices[face2[1]], vertices[face2[2]]};
+        edge = {vertices[entry.first.first], vertices[entry.first.second]};
+
+        g_dist_mat.at<float>(adjacent_faces.first, adjacent_faces.second) = getGeodesicDistance(f1, f2, edge);
+        a_dist_mat.at<float>(adjacent_faces.first, adjacent_faces.second) = getAngleDistance(f1, f2);
+    }
+
+    // normalize
+    g_dist_mat /= (cv::sum(g_dist_mat)[0] / num_adjacency);
+    a_dist_mat /= (cv::sum(a_dist_mat)[0] / num_adjacency);
+    // make symmetric
+    g_dist_mat += g_dist_mat.t();
+    a_dist_mat += a_dist_mat.t();
+    // compute output distance matrix
+    out = this->delta * g_dist_mat + (1.f - this->delta) * a_dist_mat;
+}
+
+float SpectralCluster::Impl::getGeodesicDistance(const std::vector<Point3f>& face1, const std::vector<Point3f> &face2,
+                                           const std::pair<Point3f, Point3f> &edge) {
+    cv::Point3f centroid1 = calculateFaceCentroid(face1);
+    cv::Point3f centroid2 = calculateFaceCentroid(face2);
+    cv::Point3f edge_center = (edge.first + edge.second) / 2.f;
+    return float(norm(centroid1 - edge_center)) + float(norm(centroid2 - edge_center));
+}
+
+Point3f SpectralCluster::Impl::calculateFaceCentroid(const std::vector<cv::Point3f>& face) {
+    cv::Point3f centroid(0.0f, 0.0f, 0.0f);
+    for(const cv::Point3f& p : face)
+        centroid += p;
+    return centroid / 3.f;
+}
+
+float SpectralCluster::Impl::getAngleDistance(const std::vector<Point3f>& face1, const std::vector<Point3f> &face2) const {
+    cv::Vec3f normal1 = calculateFaceNormal(face1);
+    cv::Vec3f normal2 = calculateFaceNormal(face2);
+    float cos_angle = normal1.dot(normal2);
+    float angular_distance = 1.0f - cos_angle;
+
+    if (cos_angle < 0)
+        angular_distance *= this->eta;
+    return angular_distance;
+}
+
+Point3f SpectralCluster::Impl::calculateFaceNormal(const std::vector<Point3f>& face) {
+    Vec3f normal = Vec3f(face[1] - face[0]).cross(Vec3f(face[2] - face[0]));
+    normal /= cv::norm(normal);
+    return normal;
+}
+
+void SpectralCluster::cluster(const std::vector<cv::Point3f> &vertices, const std::vector<std::vector<int32_t>> &indices,
+                              int k, OutputArray result) {
```

Comment:
Please check input parameters for emptyness
(later when they are replaced by `InputArray` instances there should be also a type check)